include/rmm/mr/default_memory_resource.hpp

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "device_memory_resource.hpp"
+namespace rmm {
+namespace mr {
+
+/**---------------------------------------------------------------------------*
+ * @brief Get the default device memory resource pointer.
+ *
+ * The default device memory resource is used when an explicit memory resource
+ * is not supplied. The initial default memory resource is a
+ * `cuda_memory_resource`.
+ *
+ * The default memory resource is used for all temporary memory allocation.
+ *
+ * This function is thread-safe.
+ *
+ * @return device_memory_resource* Pointer to the current default memory
+ * resource
+ *---------------------------------------------------------------------------**/
+device_memory_resource* get_default_resource();
+
+/**---------------------------------------------------------------------------*
+ * @brief Sets the default device memory resource pointer.
+ *
+ * If `new_resource` is not `nullptr`, sets the default device memory resource
+ * pointer to `new_resource`. Otherwise, resets the default device memory
+ * resource to the initial `cuda_memory_resource`.
+ *
+ * It is the caller's responsibility to maintain the lifetime of the object
```

Comment:
How will we manage this lifetime in libcudf, which currently has no state?

Reply:
So there's two places where I see `set_default_resource()` being called:

1. In RMM's initialization (i.e., set the default resource to `cnmem` vs. `cuda` vs. `managed`. RMM will maintain the lifetime of these objects in its state.
2. Directly by users of RMM and other RAPIDS libraries. If they do this, it's their responsibility to maintain the lifetime of their resource.

Reply:
Again, I was thinking that your proposal was to replace RMM's global state in libcudf with passing memory resources to every API.

Reply:
Here's my current thinking of what RMM will do:
- Maintain global state such as default memory resource
- Use `rmmInitialize` to set global state 
   - Or we could move initialization to happen within the first RMM call that is made, like initialization of a CUDA context on the first runtime API invoked
- Provide optional logging for allocs/free
- For the medium term, `rmm::alloc/free` will just forward calls to the default memory resource
   - Long term, these functions go away? Not sure yet.

Reply:
If `rmm::free()` has to call `mr::deallocate()`, that means RMM has to maintain a map of allocations to sizes, which I don't like.

Reply:
For all of our existing resources, the `size` argument for `deallocate` is ignored, so I don't think it'd be too much of a problem.

Reply:
Ah, that wasn't clear.